Basic/APIDiorama/Assets/UseCaseSamples/NetworkVariables/Scripts/CharacterManager.cs

```diff
@@ -0,0 +1,125 @@
+using TMPro;
+using Unity.Collections;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Unity.Netcode.Samples.APIDiorama
+{
+    /// <summary>
+    /// A complex data structure. Can only contain the types listed here: https://docs-multiplayer.unity3d.com/netcode/current/basics/networkvariable/index.html#supported-types
+    /// </summary>
+    struct SyncableCustomData : INetworkSerializable
+    {
+        public int Health;
+        public FixedString128Bytes Username; //value-type version of string with fixed allocation
+
+        public void NetworkSerialize<T>(BufferSerializer<T> serializer) where T : IReaderWriter
+        {
+            serializer.SerializeValue(ref Health);
+            serializer.SerializeValue(ref Username);
+        }
+    }
+
+    /// <summary>
+    /// Manages the data of a character
+    /// </summary>
+    public class CharacterManager : NetworkBehaviour
+    {
+        /// <summary>
+        /// The NetworkVariable holding the custom data to synchronize.
+        /// </summary>
+        NetworkVariable<SyncableCustomData> m_SyncedCustomData = new NetworkVariable<SyncableCustomData>(writePerm: NetworkVariableWritePermission.Owner); //you can adjust who can write to it with parameters
+
+        [SerializeField] Image m_HealthBarImage;
+        [SerializeField] TMP_Text m_UsernameLabel;
+
+        [SerializeField, Tooltip("The seconds that will elapse between data changes")]
+        float m_SecondsBetweenDataChanges;
+        float m_ElapsedSecondsSinceLastChange;
+
+        public override void OnNetworkSpawn()
+        {
+            base.OnNetworkSpawn();
+            if (IsClient)
+            {
+                /*
+                 * We call the color change method manually when we connect to ensure that our color is correctly initialized. 
+                 * This is helpful for when a client joins mid-game and needs to catch up with the current state of the game.
+                 */
+                OnClientCustomDataChanged(m_SyncedCustomData.Value, m_SyncedCustomData.Value);
+                m_SyncedCustomData.OnValueChanged += OnClientCustomDataChanged; //this will be called on the client whenever the value is changed by the server
+            }
+        }
+
+        public override void OnNetworkDespawn()
+        {
+            base.OnNetworkDespawn();
+            if (IsClient)
+            {
+                m_SyncedCustomData.OnValueChanged -= OnClientCustomDataChanged;
+            }
+        }
+
+        void Update()
+        {
+            if (!IsSpawned)
```

Comment:
Boss Room side, the pattern we have is to enable/disable the component on spawn and check for isServer there too. This way you remove useless update calls that always early return.

Reply:
another pattern we have is to name classes according to where they are going to execute. So here, knowing you'll always early return in Update, you could name it like ServerCharacterManager. This helps when reading that code to understand the execution context

Reply:
same 2 comments applies to other scripts doing the same things

Reply:
> On the other hand, this sample is meant to be copy-pastable. This means we should provide our users with a foundation that can be easily extended without the need of being reworked.

Updates that early return still have overhead. Enough to actually see it in the profiler (just did the test: 200 GameObjects each with 8 monobehaviour with their Update that early return shows 0.44ms)
With production games with tight frame budgets, you want to remove as much overhead as possible.
What would be the use case for an early return like this in production games? The way I see this, all use cases with an Update that never executes server or client side should be disabled, not early returned. But I might be missing the use case though.

> This script also includes client-related functionality, so calling it "Server" would be extremely misleading.

Oh I missed that, nvm then :)